contracts/AMMStrategy.sol

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Ownable } from "./abstracts/Ownable.sol";
+import { Asset } from "./libraries/Asset.sol";
+
+import { IWETH } from "./interfaces/IWeth.sol";
+import { IUniswapPermit2 } from "./interfaces/IUniswapPermit2.sol";
+import { IAMMStrategy } from "./interfaces/IAMMStrategy.sol";
+import { IStrategy } from "./interfaces/IStrategy.sol";
+
+contract AMMStrategy is IAMMStrategy, Ownable {
+    using SafeERC20 for IERC20;
+
+    address public immutable weth;
+    address public immutable permit2;
+    address public entryPoint;
+    mapping(address => bool) public ammMapping;
+
+    receive() external payable {}
+
+    /************************************************************
+     *              Constructor and init functions               *
+     *************************************************************/
+
+    constructor(
+        address _owner,
+        address _entryPoint,
+        address _weth,
+        address _permit2,
+        address[] memory _ammAddrs
+    ) Ownable(_owner) {
+        entryPoint = _entryPoint;
+        weth = _weth;
+        permit2 = _permit2;
+        for (uint256 i = 0; i < _ammAddrs.length; ++i) {
+            ammMapping[_ammAddrs[i]] = true;
+            emit SetAMM(_ammAddrs[i], true);
+        }
+    }
+
+    /************************************************************
+     *                 Internal function modifier                *
+     *************************************************************/
+    modifier onlyEntryPoint() {
+        require(msg.sender == entryPoint, "only entry point");
+        _;
+    }
+
+    /************************************************************
+     *           Management functions for Owner               *
+     *************************************************************/
+    /// @inheritdoc IAMMStrategy
+    function setAMMs(address[] calldata _ammAddrs, bool[] calldata _enables) external override onlyOwner {
```

Comment:
* AMM cannot be an ERC-20 tokens, which user can change allowance of the tokens.
* For example, if WETH is an AMM, then user can `WETH.approve(SUSHISWAP,0)`

Reply:
Treat WETH as an AMM is a cool idea that I never thought before 👍